### PR TITLE
1621275: Less communication with candlepin server from sub-man plugin; ENT-923

### PR DESCRIPTION
**Pull request by jirihnidek**
* Bug fix: https://bugzilla.redhat.com/show_bug.cgi?id=1621275
* Added several hooks to minimize communication with candlepin
  server.
* Do not try to communicate with candlepin server, when "local
  yum commands are executed (e.g. yum --help, list, check,
  history, etc.).
* The presence of identity and entitlements certificates on
  the disk of host is checked in postconfig hook
* Updating of redhat.repo is done in prereposetup_hook
* When rhsm.full_refresh_on_yum == 0, then the plugin doesn't
  communicate with candlepin server at all, because it wasn't
  necessary at all.
* Small refactoring of code

### CODE REVIEW

File: src/plugins/subscription-manager.py [hunk 1]
```diff
@@ -159,7 +159,7 @@ def warnOrGiveUsageMessage(conduit):
             conduit.info(2, msg)
 
 
-def postconfig_hook(conduit):
+def posttrans_hook(conduit):
```
Comment (nikosmoum) [inline src/plugins/subscription-manager.py]: It works, but the output is displayed in the middle of the installation which can be confusing:
```
[vagrant@centos7 vagrant]$ sudo yum install tree httpd
...
Running transaction
  Installing : httpd-2.4.6-88.el7.centos.x86_64                                                      1/2 
  Installing : tree-1.6.0-10.el7.x86_64                                                              2/2 
This system is registered with an entitlement server, but is not receiving updates. You can use subscription-manager to assign subscriptions.
  Verifying  : tree-1.6.0-10.el7.x86_64                                                              1/2 
  Verifying  : httpd-2.4.6-88.el7.centos.x86_64                                                      2/2 
...
```
It is probably cleaner to use `pretrans_hook`, which shows the display right after the `Running Transaction` message:
```
[vagrant@centos7 vagrant]$ sudo yum install tree httpd
...
Running transaction
This system is registered with an entitlement server, but is not receiving updates. You can use subscription-manager to assign subscriptions.
  Installing : apr-1.4.8-3.el7_4.1.x86_64                                                            1/6 
  Installing : apr-util-1.5.2-6.el7.x86_64                                                           2/6 
  Installing : httpd-tools-2.4.6-88.el7.centos.x86_64                                                3/6 
  Installing : mailcap-2.1.41-2.el7.noarch                                                           4/6 
  Installing : httpd-2.4.6-88.el7.centos.x86_64                                                      5/6 
  Installing : tree-1.6.0-10.el7.x86_64                                                              6/6 
  Verifying  : mailcap-2.1.41-2.el7.noarch                                                           1/6 
  Verifying  : httpd-tools-2.4.6-88.el7.centos.x86_64                                                2/6 
  Verifying  : apr-util-1.5.2-6.el7.x86_64                                                           3/6 
  Verifying  : tree-1.6.0-10.el7.x86_64                                                              4/6 
  Verifying  : httpd-2.4.6-88.el7.centos.x86_64                                                      5/6 
  Verifying  : apr-1.4.8-3.el7_4.1.x86_64                                                            6/6 
                                               1/1 
...
```

Comment (jirihnidek) [reply src/plugins/subscription-manager.py]: I tested my change (and your proposal) on rhel and the change could not be so simple. Why? When you use RHEL, then all repositories are defined in redhat.repo. When the host is unregistered, then you cannot install any package and you will not see the message: `"This system is not registered with an entitlement server..."` I have some idea, but I would like to discuss it with the rest of the team.

**Comment by nikosmoum** [comment]
@candlepin-pull-request-bot retest this please

**Comment by jirihnidek** [comment]
> But the commands list, repolist and info, produce as many requests as the the others (provides, install, groups, deplist, search commands):

Commands `list`, `repolist` and `info` read content of repository. Thus hook `prereposetup_hook` is triggered and CP requests are produced. We cannot do much about it.

**Pull request merged**